Enable localization of read/search page with url param

roo2 opened the pull request:
Adds support for localization to the logged out read/search page so that it can be a suitable replacement for the public search page.

The approach recomemnded in pe7F0s-ZH-p2#comment-1126 was to use the same scheme as existing public pages such as `/themes` which includes the language code in the URL e.g. `/fr/themes`.

I was not able to use the exact same approach as those pages because they use the `ssrSetupLocale` middleware which only works for server side rendered pages, however the search page does not currently support server side rendering.

This approach works on the client however using the existing `setLocaleMiddleware`. The `lang` param is also passed along to the search endpoint. This will also have the affect of localizing other reader stream endpoints. I also updated the search suggestions to be localized if the user is logged out.


### Testing instructions

In an incognito window go to a localized route such as /fr/read/search
Search results should also be localized.

roo2 committed:
draft: disable filters to test translation

roo2 committed:
try make search isomorphic

roo2 committed:
fix

roo2 committed:
fix

roo2 committed:
use setlocale middleware on search page

roo2 committed:
use lang param for search and reader API calls

roo2 committed:
fall back to context.lang if locale isn't set for search suggestions

davemart-in commented:
One minor thing:

If I type in `/fr/read/search` in the URL, the header is not translated:

<img width="1436" alt="CleanShot 2023-06-27 at 09 51 10@2x" src="https://github.com/Automattic/wp-calypso/assets/5634774/32c54378-d2bc-46fd-b992-b9c3e8b40a2f">

But if I click a search link, or click the back button it is translated:

https://github.com/Automattic/wp-calypso/assets/5634774/2b517a58-ff8e-41ae-9a6c-f2517ff08636

creativecoder commented on client/sections.js:
```diff
@@ -360,7 +360,7 @@ const sections = [
 	{
 		name: 'reader',
-		paths: [ '/read/search', '/recommendations' ],
+		paths: [ '/read/search', '/([a-z]{2,3}|[a-z]{2}-[a-z]{2})/read/search', '/recommendations' ],
```
Just curious, what are the 3 letter language codes that can be used?

roo2 replied:
It looks like we have a few, "als" for Alemannisch (the language of the Amish)  https://github.com/Automattic/wp-calypso/blob/trunk/packages/languages/src/languages-meta.json#L12

creativecoder reviewed (commented):
This seem like a reasonable way to handle this, given the constraints--nice work!

Works as described for me, except for the header translation issue that @lessbloat noted.

Another issue I noticed when logged out is that the sidebar seems to be trying to show a subscribed sites list, which is empty:

<img width="1099" alt="image" src="https://github.com/Automattic/wp-calypso/assets/1699996/06c1e261-256f-43ce-98b1-888c79430307">

roo2 committed:
use regex to test routes for /read/search pattern

roo2 committed:
tmp: add logging to debug mistranslted header

roo2 committed:
fix regex

roo2 committed:
use a function to render header section so that re-render is triggered when locale data becomes available

roo2 commented on client/layout/logged-out.jsx:
```diff
@@ -161,7 +167,9 @@ const LayoutLoggedOut = ( {
 				{ masterbar }
-				{ headerSection && <div className="layout__header-section-content">{ headerSection }</div> }
+				{ renderHeaderSection && (
+					<div className="layout__header-section-content">{ renderHeaderSection() }</div>
```
I had to change renderHeaderSection to a function because of how setting locale works on the front end:
* When setLocale is called, the client asynchronously requests a translation file for that locale.
* When the XHR for that completes, a state change event occurs which triggers a re-render.
* But if `headerSection` is passed in as a pre-rendered FC it will not be re-rendered with the new translations.

roo2 commented:
I fixed those two things:
* Changed to use regex to check for whether to render the "related sites" or "popular sites"
  * This may conflict with the changes you're making @creativecoder? I'll make sure to rebase and retest once your changes are in.
* Change to not pre-render the header section before the translations are in.
  * I made it pass in a render function rather than the component, that way re-rendering is triggered after the locale translation file is loaded (translations file is loaded asynchronously).

roo2 committed:
fix tests

davemart-in commented:
Last super minor (but still likely important) thing:

Can we update the links under the search box to contain the language specific URL?

![CleanShot 2023-06-29 at 07 48 53@2x](https://github.com/Automattic/wp-calypso/assets/5634774/b13574e7-01cf-4643-b96e-69c0c3df7cfb)

Else, when they click there, they'll always see the English results.

yuliyan commented on client/blocks/reader-post-card/index.jsx:
```diff
@@ -143,7 +144,11 @@ class ReaderPostCard extends Component {
 		const isConversations = currentRoute.startsWith( '/read/conversations' );
-		const isReaderSearchPage = currentRoute.startsWith( '/read/search' );
+
+		const isReaderSearchPage = !! new RegExp( `^(/${ localeRegexString })?/read/search` ).test(
```
Does the `!!` have any purpose here?

roo2 replied:
Ahh no it's not necessary, previously it was using `match` which returns an array or null, but `test` already returns a bool 👍. (the component that accepts it strictly wants a bool param)

yuliyan reviewed (commented):
A few things that I noticed:

1. The suggestion links are not localized. Using the links would initially render the translated screen, but after refreshing the page it would render the English pages.
![CleanShot 2023-06-29 at 16 00 25](https://github.com/Automattic/wp-calypso/assets/2722412/488cab27-cad9-4179-a39f-bdca2db32cc1)

2. It seems like `/{lang}/read/search` works for logged in users as well, but I think it would make more sense to redirect them to `/read/search` so that it uses the user preferences language.

roo2 committed:
localise search suggestion links

roo2 committed:
redirect to default locale if user logged in

roo2 commented:
I've fixed the issues raised:
* Suggestion links now localized
* logged in users are redirected to the non-localized route
* tidy up code

davemart-in reviewed (approved):
I'd say ship it. It's already much better than what we have today. Getting translations perfect is going to be difficult. We can always keep making small iterative improvements down the road.
